### PR TITLE
Prohibit multiple outgoing connections to the same peer.

### PR DESCRIPTION
**Pull request by fyookball**
I tested it but was unable to actually see a case when there were duplicates, so not all code paths fired.

### CODE REVIEW

**Comment by fyookball** [comment]
This is for issue #142

File: connmgr/connmanager.go [hunk 1]
```diff
@@ -35,8 +35,39 @@ var (
...
 
+// concurrentConnectionMap is safe to use concurrently.
+type concurrentConnectionMap struct {
+	connections map[string]bool
+	mux         sync.Mutex
+}
+
+// addToConnectionMap adds a network address to the map.
+func (c *concurrentConnectionMap) addToConnectionMap(key string) {
+	c.mux.Lock()
```
Comment (zquestz) [inline connmgr/connmanager.go]: NIT:

```
c.mux.Lock()
defer c.mux.Unlock()
c.connections[key] = true
```

File: connmgr/connmanager.go [hunk 2]
```diff
@@ -35,8 +35,39 @@ var (
...
+	delete(c.connections, key)
+	c.mux.Unlock()
+}
+
+//findKeyInConnectionMap returns a boolean of whether a network address is in the map.
+func (c *concurrentConnectionMap) findKeyInConnectionMap(key string) bool {
+	c.mux.Lock()
+	_, found := c.connections[key]
+	defer c.mux.Unlock()
```
Comment (zquestz) [inline connmgr/connmanager.go]: Move this up a line =)

File: connmgr/connmanager.go [hunk 3]
```diff
@@ -263,6 +294,9 @@ out:
 				connReq.updateState(ConnEstablished)
 				connReq.conn = msg.conn
 				conns[connReq.id] = connReq
+				// Add the address to the peerMap
+				peerNetworkAddressString := connReq.Addr.String()
```
Comment (zquestz) [inline connmgr/connmanager.go]: NIT: No need to suffix everything with their type =)

**Comment by fyookball** [comment]
@zquestz thanks. Nits cleaned.

**Comment by cpacia** [comment]
So a couple things I see..

Looking at the existing code, it doesn't look like a failed connection request ever gets deleted from the `pending` map. I'm not sure that is by design. Can anyone think of any reason why we would need to leave a connect request for a failed connection in the pending map?

I bring this up because the way the code is structured where it just returns from the `connect` function without doing anything else will leave the connect request sitting in memory. 

And just on another stylistic point. This can probably be accomplished without a mutex by maybe having the `registerPending` case return a bool over the done channel for whether or not the ip address of the request exists in either `conns` map or the `pending` map. 

If it does exist `NewConnReq` should also call `handleFailedConn()` so that it ill try another address.

**Comment by fyookball** [comment]
I reworked this code per @cpacia main idea here.   In connect(), we aren't fetching an address to specify to the connection manager, so even though it passes the duplicateAddress channel in Connect(), it doesnt make use of it; I assume this ok.

File: connmgr/connmanager.go [hunk 4]
```diff
@@ -369,27 +383,33 @@ func (cm *ConnManager) NewConnReq() {
...
+	case <-duplicateAddress:
+		duplicated = true
 	case <-done:
 	case <-cm.quit:
 		return
 	}
 
-	addr, err := cm.cfg.GetNewAddress()
-	if err != nil {
+	//addr, err := cm.cfg.GetNewAddress()
```
Comment (zquestz) [inline connmgr/connmanager.go]: NIT: Can remove =)

File: connmgr/connmanager.go [hunk 5]
```diff
@@ -246,6 +248,18 @@ out:
 				connReq := msg.c
 				connReq.updateState(ConnPending)
 				pending[msg.c.id] = connReq
+				// check duplicates
+				var duplicatesFound bool = false
```
Comment (zquestz) [inline connmgr/connmanager.go]: NIT: `var duplicatesFound = false`

File: connmgr/connmanager.go [hunk 6]
```diff
@@ -246,6 +248,18 @@ out:
...
+
+				for _, v := range conns {
+					if v.Addr.String() == msg.networkAddress {
+						duplicatesFound = true
+					}
+				}
+				if duplicatesFound == true {
```
Comment (zquestz) [inline connmgr/connmanager.go]: `if duplicatesFound`

File: connmgr/connmanager.go [hunk 7]
```diff
@@ -369,27 +383,33 @@ func (cm *ConnManager) NewConnReq() {
...
 	done := make(chan struct{})
+	duplicateAddress := make(chan bool)
 	select {
-	case cm.requests <- registerPending{c, done}:
+	case cm.requests <- registerPending{c, done, addr.String(), duplicateAddress}:
 	case <-cm.quit:
 		return
 	}
 
+	var duplicated bool = false
```
Comment (zquestz) [inline connmgr/connmanager.go]: NIT: `var duplicated = false`

**Comment by zkvm** [comment]
> Looking at the existing code, it doesn't look like a failed connection request ever gets deleted from the `pending` map. I'm not sure that is by design. Can anyone think of any reason why we would need to leave a connect request for a failed connection in the pending map?

A failed connection could get connected after a retry. Caller can explicitly use `Remove`  method to cancel a failed connect request and it will be deleted from pending map

**Comment by zquestz** [comment]
New version of this was created. Closing.

**Pull request closed without merging**